[feat] Add support for the OAR scheduler

mahendrapaipuri opened the pull request:
This PR adds support to a minimal [OAR](https://oar.imag.fr/documentation) backend scheduler. 
Fixes #2124.

teojgo commented:
Ok to test

teojgo commented:
@jenkins-cscs retry all

vkarak commented on reframe/core/schedulers/oar.py:
```diff
@@ -0,0 +1,201 @@
+                preamble.append(self._format_option(opt))
+
+        # OAR starts the job in the home directory by default
+        preamble.append('cd %s' % job.workdir)
```
Use f-strings here.

mahendrapaipuri replied:
Fixed

vkarak commented on reframe/core/schedulers/oar.py:
```diff
@@ -0,0 +1,201 @@
+            return
+
+        for job in jobs:
+            completed = osext.run_command(
+                f'oarstat -fj {job.jobid}'
+            )
+
+            if completed.returncode != 0:
+                raise JobSchedulerError(
+                    f'oarstat failed with exit code {completed.returncode} '
+                    f'(standard error follows):\n{completed.stderr}'
+                )
```
Just use `_run_strict()` here and do not throw an extra error.

mahendrapaipuri replied:
Fixed

vkarak reviewed (changes requested):
Thanks @mahendrapaipuri for your PR I cannot test the backend, so my comments are related to the code itself.

vkarak commented:
We would also need an entry in the documentation.

mahendrapaipuri commented on unittests/test_schedulers.py:
```diff
@@ -208,6 +209,22 @@ def _expected_torque_directives(job):
+        '#OAR -O %s' % job.stdout,
+        '#OAR -E %s' % job.stderr,
+        '#OAR -l /host=%s/core=%s,walltime=0:5:0' % (num_nodes,
+                                                     num_tasks_per_node),
```
Fixed

mahendrapaipuri commented on reframe/core/schedulers/oar.py:
```diff
@@ -0,0 +1,201 @@
+            self._format_option('-O %s' %
+                                os.path.join(job.workdir, job.stdout)),
+            self._format_option('-E %s' %
+                                os.path.join(job.workdir, job.stderr)),
```
Fixed

mahendrapaipuri commented:
> We would also need an entry in the documentation.

Added an entry in the documentation. I added a line about CI image as well. I am sorry that I forgot to push that change in that previous PR.

vkarak commented on docs/tutorial_tips_tricks.rst:
```diff
@@ -552,8 +552,10 @@ The following is an example of ``.gitlab-ci.yml`` file that does exactly that:
-We then pass the generated CI pipeline file to second phase as an artifact and we are done!
+ReFrame will process them as usual, but instead of running the selected tests, it will generate the correct steps
+for running each test individually as a Gitlab job. We then pass the generated CI pipeline file to second phase as
+an artifact and we are done! If ``image`` keyword is defined in ``.gitlab-ci.yml``, the emitted pipeline will use
+the same image as the one defined in the parent pipeline.
```
We have a policy in the documentation to start new sentences in new lines for making diffs easier:

```suggestion
an artifact and we are done! 
If the ``image`` keyword is defined in ``.gitlab-ci.yml``, the emitted pipeline will use
the same image as the one defined in the parent pipeline.
The verbosity level of the initial invocation of ReFrame will also be retained in the invocations from the child pipeline.
```

vkarak replied:
It'd be better if we had this doc enhancement in a separate PR.

mahendrapaipuri replied:
Ok. Will do that in the future PRs.

vkarak reviewed (changes requested):
These changes broke the unit tests for `test_prepare[oar]`:

```
E       assert {'#DW jobdw capacity=100GB',\n '#DW stage_in source=/foo',\n '#OAR --gres=gpu:4',\n '#OAR -E '\n '/private/var/folders/h7/k7cgrdl13r996m4dmsvjq7v80000gp/T/pytest-of-karakasv/pytest-111/test_prepare_oar_0/job.err',\n '#OAR -O '\n '/private/var/folders/h7/k7cgrdl13r996m4dmsvjq7v80000gp/T/pytest-of-karakasv/pytest-111/test_prepare_oar_0/job.out',\n '#OAR -l /host=8/core=2,walltime=0:5:0#OAR --account=spam',\n '#OAR -n "testjob"'} == {'#DW jobdw capacity=100GB',\n '#DW stage_in source=/foo',\n '#OAR --account=spam',\n '#OAR --gres=gpu:4',\n '#OAR -E '\n '/private/var/folders/h7/k7cgrdl13r996m4dmsvjq7v80000gp/T/pytest-of-karakasv/pytest-111/test_prepare_oar_0/job.err',\n '#OAR -O '\n '/private/var/folders/h7/k7cgrdl13r996m4dmsvjq7v80000gp/T/pytest-of-karakasv/pytest-111/test_prepare_oar_0/job.out',\n '#OAR -l /host=8/core=2,walltime=0:5:0',\n '#OAR -n "testjob"'}
E         Extra items in the left set:
E         '#OAR -l /host=8/core=2,walltime=0:5:0#OAR --account=spam'
E         Extra items in the right set:
E         '#OAR -l /host=8/core=2,walltime=0:5:0'
E         '#OAR --account=spam'
E         Full diff:
E           {
E            '#DW jobdw capacity=100GB',
E            '#DW stage_in source=/foo',
E         -  '#OAR --account=spam',
E            '#OAR --gres=gpu:4',
E            '#OAR -E '
E            '/private/var/folders/h7/k7cgrdl13r996m4dmsvjq7v80000gp/T/pytest-of-karakasv/pytest-111/test_prepare_oar_0/job.err',
E            '#OAR -O '
E            '/private/var/folders/h7/k7cgrdl13r996m4dmsvjq7v80000gp/T/pytest-of-karakasv/pytest-111/test_prepare_oar_0/job.out',
E         -  '#OAR -l /host=8/core=2,walltime=0:5:0',
E         +  '#OAR -l /host=8/core=2,walltime=0:5:0#OAR --account=spam',
E         ?                                        +++++++++++++++++++
E            '#OAR -n "testjob"',
E           }

```

vkarak reviewed (approved):
Lgtm now. I fixed the unit tests.

mahendrapaipuri commented:
Thanks @vkarak for fixing unit tests.
